src/client/tile.cpp

```diff
@@ -1760,9 +1760,46 @@ bool TextureSource::generateImagePart(std::string part_of_name,
 			for (u32 y = 0; y < dim.Height; y++)
 			for (u32 x = 0; x < dim.Width; x++)
 			{
-				video::SColor c = baseimg->getPixel(x,y);
+				video::SColor c = baseimg->getPixel(x, y);
 				c.setAlpha(floor((c.getAlpha() * ratio) / 255 + 0.5));
-				baseimg->setPixel(x,y,c);
+				baseimg->setPixel(x, y, c);
+			}
+		}
+		/*
+			[invert:mode
+			Inverts the given channels of the base image.
+			Mode may contain the characters "r", "g", "b", "a".
+			Only the channels that are mentioned in the mode string
+			will be inverted.
+		*/
+		else if (str_starts_with(part_of_name, "[invert:")) {
+			if (baseimg == NULL) {
+				errorstream << "generateImagePart(): baseimg == NULL "
+						<< "for part_of_name=\"" << part_of_name
+						<< "\", cancelling." << std::endl;
+				return false;
+			}
+
+			Strfnd sf(part_of_name);
+			sf.next(":");
+
+			std::string mode = sf.next("");
+			bool r = mode.find("r") != std::string::npos;
+			bool g = mode.find("g") != std::string::npos;
+			bool b = mode.find("b") != std::string::npos;
+			bool a = mode.find("a") != std::string::npos;
+
+			core::dimension2d<u32> dim = baseimg->getDimension();
+
+			for (u32 y = 0; y < dim.Height; y++)
+			for (u32 x = 0; x < dim.Width; x++)
+			{
+				video::SColor c = baseimg->getPixel(x, y);
+				if (r) c.setRed(255 - c.getRed());
+				if (g) c.setGreen(255 - c.getGreen());
+				if (b) c.setBlue(255 - c.getBlue());
+				if (a) c.setAlpha(255 - c.getAlpha());		
+				baseimg->setPixel(x, y, c);
```

Comment:
This loop seems inefficient to me (though I could be wrong; the compiler might optimize it if it is really clever, but I doubt it). Avoid having branches inside loops with many iterations.

My suggestion is based around the fact that `255 - x` is equivalent to `x ^ 255` (where ^ is XOR) if `x` is an integer between 0 and 255. Similarly, XORing with 0 leaves `x` unchanged. So, prior to the loop, we can compute an XOR mask for each channel which is 255 if that channel should be inverted and 0 if it should not.

Also, instead of having one XOR mask per channel, a single 32-bit mask can be used. This saves the need to call getRed, setRed, (etc. for the other channels) which internally work by masking and shifting some bits anyway. All that is extra work that is not necessary.

So, the overall algorithm would look like this (pseudocode):
- Initialize `u32 mask = 0`
- If alpha should be inverted, then do `mask |= 0xff000000UL`
- If red should be inverted, then do `mask |= 0x00ff0000UL`
- If green should be inverted, then do `mask |= 0x0000ff00UL`
- If blue should be inverted, then do `mask |= 0x000000ffUL`
- For each pixel x,y in the image, do:
  - `setPixel(x, y, getPixel(x, y) ^ mask)`.

For optimum efficiency, the setPixel and getPixel calls could be replaced by locking and working directly with the image data, but that could lead to problems if the image does not have the right color format, so it might not be worth it.